core/src/main/java/co/cask/wrangler/utils/Json2Schema.java

```diff
@@ -0,0 +1,202 @@
+ */
+public final class Json2Schema {
+  // Parser to parse json if provided as string.
+  private final JsonParser parser;
+
+  public Json2Schema() {
+    this.parser = new JsonParser();
+  }
+
+  /**
+   * Generates a {@link Schema} given a record.
+   *
+   * @param id Schema id
+   * @return returns {@link Schema}
+   * @throws UnsupportedTypeException
+   * @throws JSONException
+   */
+  public Schema toSchema(String id, Record record) throws RecordConvertorException {
+    List<Schema.Field> fields = new ArrayList<>();
+
+    // Iterate through each field in the record.
+    for (KeyValue<String, Object> column : record.getFields()) {
+      String name = column.getKey();
+      Object value = column.getValue();
+
+      // First, we check if object is of simple type.
+      if (value instanceof String || value instanceof Integer || value instanceof Long || value instanceof Short ||
+          value instanceof Double || value instanceof Float || value instanceof Boolean || value instanceof byte[]) {
+        try {
+          Schema schema = Schema.nullableOf(new SimpleSchemaGenerator().generate(value.getClass()));
+          fields.add(Schema.Field.of(name, schema));
+        } catch (UnsupportedTypeException e) {
+          throw new RecordConvertorException(
+            String.format("Unable to convert field '%s' to basic type.", name)
+          );
+        }
+      }
+
+      if (value instanceof Map) {
+        Schema schema = Schema.nullableOf (Schema.mapOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.STRING)));
+        fields.add(Schema.Field.of(name, schema));
+      }
+
+      if (value instanceof JsonElement) {
+        Schema schema = toSchema(name, (JsonElement) value);
+        fields.add(Schema.Field.of(name, schema));
+      }
+    }
+    return Schema.recordOf(id, fields);
+  }
+
+  /**
+   * Converts a json string to a {@link Schema}.
+   *
+   * @param id of the {@link Schema}
+   * @param json represents the JSON as {@link String}
+   * @return instance of {@link Schema} representing the JSON.
+   * @throws RecordConvertorException throw if there are any issues parsing to {@link Schema}
+   * @see {@link #toSchema(String, Record)}
+   */
+  public Schema toSchema(String id, String json) throws RecordConvertorException {
+    JsonElement element = parser.parse(json);
+    return toSchema(id, element);
+  }
+
+  private Schema toSchema(String name, JsonElement element) throws RecordConvertorException {
+    Schema schema = null;
+    if (element.isJsonObject() || element.isJsonArray()) {
+      schema = toComplexSchema(name, element);
+    } else if (element.isJsonPrimitive()) {
+      schema = toSchema(name, element.getAsJsonPrimitive());
+    } else if (element.isJsonNull()) {
+      schema = Schema.of(Schema.Type.NULL);
+    }
+    return schema;
+  }
+
+  private Schema toSchema(String name, JsonArray array) throws RecordConvertorException {
+    int[] types = new int[3];
```

Comment:
would it be easier to read to only have two booleans instead of this array? ``hasComplexTypes`` and ``hasSimpleTypes``?